### ISSUE TITLE
Language is switching repeatedly  on network failure simulate

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#13521]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open setting page.
2. Go to preference.
3. Switch language to another and switch back to previous one.
4. Stay on Preference page and wait for few seconds.

## Expected Result:
Shouldn’t switch repeatedly and last selected language should be applied.

## Actual Result:
Language is switching repeatedly.

## Workaround:
unknown

## Platform:

Where is this issue occurring?

- iOS

**Version Number:** 1.2.38-0
**Reproducible in staging?:** y
**Reproducible in production?:**  Toggle is not in production
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/207074281-ded66068-5177-47d6-8378-aa51cecf1d4b.mov

https://user-images.githubusercontent.com/43996225/207074387-e9a20342-7961-4423-9fee-f7f34402064b.MP4

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @jatinsoni
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1670659690749749

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by hungvu193** [comment]
<b>Proposal</b>
Root clause: When switching between 2 languages, 2 API called at the same time and the api property `canRetry` is always true, so when 2 API requests were calling and retrying at the same time. As I can see it retried 10 times, so you can see 1 API retry to set language to English, the another one tried to set language to Espanol. That's why you can see it's switching repeatedly.
<b>Solution </b>
If this problem only happens with ios, we can fix that by using the `onDonePress`, we can only change the language once `onDonePress` is fired, we can also add a loading while changing language since changing App Locale will need time to change.
Option 1:
Update the current `LocalPicker.js` for separate logic between IOS and other platform:
```diff
diff --git a/src/components/LocalePicker.js b/src/components/LocalePicker.js
index aafb06ea6..0119016d1 100644
--- a/src/components/LocalePicker.js
+++ b/src/components/LocalePicker.js
@@ -2,6 +2,7 @@ import _ from 'underscore';
 import React from 'react';
 import {withOnyx} from 'react-native-onyx';
 import PropTypes from 'prop-types';
+import {Platform} from 'react-native';
 import compose from '../libs/compose';
 import * as App from '../libs/actions/App';
 import withLocalize, {withLocalizePropTypes} from './withLocalize';
@@ -42,15 +43,23 @@ const localesToLanguages = {
 };
 
 const LocalePicker = (props) => {
+    const [currentLocale, setCurrentLocale] = React.useState(props.preferredLocale);
     if (!Permissions.canUseInternationalization(props.betas)) {
         return null;
     }
 
     return (
         <Picker
+            onDonePress={() => {
+                if (currentLocale === props.preferredLocale || !currentLocale) {
+                    return;
+                }
+                App.setLocale(currentLocale);
+            }}
             label={props.size === 'normal' ? props.translate('preferencesPage.language') : null}
             onInputChange={(locale) => {
-                if (locale === props.preferredLocale) {
+                setCurrentLocale(locale);
+                if (locale === props.preferredLocale || Platform.OS === 'ios') {
                     return;
                 }
 
@@ -58,7 +67,7 @@ const LocalePicker = (props) => {
             }}
             items={_.values(localesToLanguages)}
             size={props.size}
-            value={props.preferredLocale}
+            value={Platform.OS === 'ios' ? currentLocale : props.preferredLocale}
         />
     );
 };
diff --git a/src/components/Picker/index.js b/src/components/Picker/index.js
index 62cb9c2b4..2c613ed41 100644
--- a/src/components/Picker/index.js
+++ b/src/components/Picker/index.js
@@ -191,6 +191,8 @@ class Picker extends PureComponent {
                             }
                             this.props.innerRef(el);
                         }}
+                        // eslint-disable-next-line react/jsx-props-no-spreading
+                        {...this.props}
                     />
                 </View>
                 <FormHelpMessage message={this.props.errorText} />

```
Option 2:
Create `LocalPicker.ios.js` and implement the logic similar with above code but only logic for IOS.

**Comment by jatinsonijs** [comment]
**Proposal**

**Reason:**  Repeatedly switching happen due to **retry cycle.** Haven't found any existing way to cancel and remove previous request and replace with new. Generally there are 2 types of request one is **Take Every** and another one is **Take Latest**. So this type of requests are take latest type in which old request will cancel and new one apply. Currently there is no direct way to do this. 

We can add request config object in write function and pass it down to request. in persistRequests we can check if request is takeLatest type then we can remove old one.

```diff
diff --git a/src/libs/API.js b/src/libs/API.js
index d86fc099a0..1426fdf250 100644
--- a/src/libs/API.js
+++ b/src/libs/API.js
@@ -17,8 +17,11 @@ import CONST from '../CONST';
  * @param {Object} [onyxData.optimisticData] - Onyx instructions that will be passed to Onyx.update() before the request is made.
  * @param {Object} [onyxData.successData] - Onyx instructions that will be passed to Onyx.update() when the response has jsonCode === 200.
  * @param {Object} [onyxData.failureData] - Onyx instructions that will be passed to Onyx.update() when the response has jsonCode !== 200.
+ * @param {Object} requestConfig - Request configuration will be use to identify
+ *                                 request behavior (cancelable, shouldRetry, takeLatest)
+ * @param {boolean} [requestConfig.takeLatest]
  */
-function write(command, apiCommandParameters = {}, onyxData = {}) {
+function write(command, apiCommandParameters = {}, onyxData = {}, requestConfig = {}) {
     // Optimistically update Onyx
     if (onyxData.optimisticData) {
         Onyx.update(onyxData.optimisticData);
@@ -34,6 +37,7 @@ function write(command, apiCommandParameters = {}, onyxData = {}) {
     // Assemble all the request data we'll be storing in the queue
     const request = {
         command,
+        ...requestConfig,
         data: {
             ...data,
 
diff --git a/src/libs/actions/App.js b/src/libs/actions/App.js
index 1e3afedb35..d357397699 100644
--- a/src/libs/actions/App.js
+++ b/src/libs/actions/App.js
@@ -85,7 +85,7 @@ function setLocale(locale) {
 
     API.write('UpdatePreferredLocale', {
         value: locale,
-    }, {optimisticData});
+    }, {optimisticData}, {takeLatest: true});
 }
 
 function setSidebarLoaded() {
diff --git a/src/libs/actions/PersistedRequests.js b/src/libs/actions/PersistedRequests.js
index 260526c13b..549b9f68d7 100644
--- a/src/libs/actions/PersistedRequests.js
+++ b/src/libs/actions/PersistedRequests.js
@@ -21,6 +21,9 @@ function clear() {
  * @param {Array} requestsToPersist
  */
 function save(requestsToPersist) {
+    if (_.first(requestsToPersist).takeLatest) {
+        persistedRequests = _.filter(persistedRequests, request => request.command !== _.first(requestsToPersist).command);
+    }
     persistedRequests = lodashUnionWith(persistedRequests, requestsToPersist, _.isEqual);
     Onyx.set(ONYXKEYS.PERSISTED_REQUESTS, persistedRequests);
 }

```

**Note:** This issue is not just about iOS and language change everywhere where only latest request should apply creating this issue. example for pin/unpin.

**Comment by Santhosh-Sellavel** [comment]
Will be reviewing proposals later today.

**Comment by jatinsonijs** [comment]
@Santhosh-Sellavel turn off network failure simulate and refresh app. once app loaded turn on network failure simulate and try to change language.

**Comment by Santhosh-Sellavel** [comment]
Thanks, @jatinsonijs!

**Comment by Santhosh-Sellavel** [comment]
@deetergp 

@s77rt [proposal](https://github.com/Expensify/App/issues/13521#issuecomment-1351145889) makes sense to me, please let me know your thoughts!

> If there is no reason to actually send the request and treat it as failed, we can just treat it as failed without sending the request. <-- Alternative proposal 3

This makes sense too!

cc: @neil-marcellini As you implemented the preference, Any thoughts?

**Comment by jatinsonijs** [comment]
This issue can happen rarely in real case. But if it is happened will create wired behaviour as shown in video.
and I don't think disabling the pusher event will solve it. as network failure simulate toggle only for testing purpose. It will be not exist in production.
This can happen request send to server => server trigger pusher event => request failed for some reason. So if it can happen in real case you should consider my proposal. but if cannot happen in real case then I think its not needed.

**Comment by Santhosh-Sellavel** [comment]
@jatinsonijs Here the problem is caused because the Request was a success -> We received a pusher event and we treated it as Failed even though the request was succeded. But in real-time if the request failed we don't get a pusher notification I expect things would work normally.

> This can happen request send to server => server trigger pusher event => request failed for some reason.

I believe that's not possible if that happened so we don't have anything to do and client-side!

**Comment by bernhardoj** [comment]
You can reproduce this even without the Simulate failing network request turned on. If you have it turned on, turned it off then reopen the app. Try to switch it (any preferences) repeatedly quickly. I explained the problem [here](https://github.com/Expensify/App/issues/13521#issuecomment-1348443012). It is easier to notice when the network fail simulate is on because as the others explained, the request is actually success, but we treat it as an error, therefore the request is retrying max 10 times "successfully", so we receive 10 Pusher API updates.

You can try to remove the Pusher subscription and the issue will be gone. (I'm not saying we need to remove it as the solution)

I don't put proposal title because I believe it need to be worked internally.

Here is my reproduction btw

https://user-images.githubusercontent.com/50919443/207756462-ff13ced3-b8cd-4945-87e5-1d2c1e3686b5.mp4

**Comment by s77rt** [comment]
@Santhosh-Sellavel 
> I am unable to reproduce this consistently

That's okay, this is due to the response timing, we are sending requests to change the language to either `en` or `es` and sometimes we get the responses in the "right" sequence that you don't notice the frequent switch i.e.: `en -> en -> en -> es -> es -> es `Other times the response timing is `en -> es -> en -> es -> en -> es`

___

@jatinsonijs @Santhosh-Sellavel 
The scenario @jatinsonijs described may actually happen since the pusher ws connection and the requests are independent. However if the request failed that's probably because the client just went offline so he won't receive the pusher notification anyway. Even if he did, it's not a problem because it will just make the change once (no rapid switching behaviour)

___

<details>
<summary markdown="span">@jatinsonijs Your solution will hide the bug not fix it and it has some issues.</summary>

1. If we remove the `optimisticData` feature and change the language while failure simulation is on, the language should not be changed.

2. You are relying on `persistedRequests` which may be empty i.e. the request has been already sent.
</details>

___

@bernhardoj This is a slightly different issue and not that critical as it won't lead to that annoying rapid switching, it will switch a **predicable** n times based on your toggles events (2 switches at max in this case). If you double-click the toggle button:
1. `optimisticData` will set the value to `off`
2. Initiate a request (A) to set that value
3. `optimisticData` will set the value to `on`
4. Initiate a request (B) to set that value
5. Receive a pusher notification (initiated by A) and  set the value to `off`
6. Receive a pusher notification (initiated by B) and  set the value to `on`

<details>
<summary markdown="span">This can be fixed but it's more or less complex that I feel it's not worth it</summary>

1. With every request add two fields `idempotent` and `timestamp`
2. Have the pusher return those two fields with relevant notifications
3. Before applying any Onyx update, check if `idempotent` is true, if it's then only apply the update if the `timestamp` is greater than the current stored one
</details>

**Comment by s77rt** [comment]
@jatinsonijs 
1. I didn't say you did, I was just trying to show you where the hidden bug is
2. Still, this does not cover the case where the request is sent already

**Comment by jatinsonijs** [comment]
Yes @s77rt  for already sent request its need to be discuss with backend for best approach. for frontend we can do thinks like debounce or cancel request if still pending etc...

**Comment by bernhardoj** [comment]
@s77rt It's totally not a different issue. What I'm explaining is the root cause of the issue. You already explained it yourself in steps that when we switch on once and then off once, it will result in on twice and off twice. But why it keeps switching multiple times when we turn on simulate network fail? It is because it keeps retrying successfully and we keep receiving the Pusher update 10 times.

What if the user has a low connection and he toggle the switch 100 times? The `optimisticData` will apply the state correctly, but then after the request success, the Pusher will switch it 100 times again, which will be the same case as you turn on the simulate network fail.

**Comment by s77rt** [comment]
@bernhardoj It's similar but not the same

> Why it keeps switching multiple times when we turn on simulate network fail?

Because we are changing the onyx value multiple times. Simple as that. Nothing is wrong here.
The issue is the fact that we are changing the value in the first place where we should not.

> What if the user has a low connection and he toggle the switch 100 times? ... the Pusher will switch it 100 times again, which will be the same case as you turn on the simulate network fail.

This is independent of the connection, the max is always 2

___

### Bugs

1. The bug here is due to applying a "nonexistent" notification
2. The bug you are referring to is due to applying an outdated notification

#### "nonexistent" notification
Notifications that we get from the pusher should not exist because the requests "failed"

#### outdated notification
Notifications that replace a new value with an old value are considered outdated

 ### Solutions
1. https://github.com/Expensify/App/issues/13521#issuecomment-1351145889
2. https://github.com/Expensify/App/issues/13521#issuecomment-1352998536

___

I think it will help to understand the issue here by "fixing" bug 2. Comment this line: https://github.com/Expensify/App/blob/a414d1a30c1726ae6dcfd44d4f28e1ccaf3b7cdf/src/libs/API.js#L24

**Comment by bernhardoj** [comment]
The issue here is because we update the state (by calling `Onyx.update`) from Pusher subscription which we should not receive. When you turn on the simulate network fail, reopen the app, and then switch any preferences repeatedly, why we can't reproduce it? (Ref https://github.com/Expensify/App/issues/13521#issuecomment-1352401524). Because the Pusher failed to subscribe (haven't check why it failed), therefore it does not receive any API update and does not call `Onyx.update()`.

**Comment by marcaaron** [comment]
If this is not affecting production then I agree it seems like a low priority item. I'm actually not too sure how useful the "simulate failing network requests" thing is. So yes, maybe we can just remove it? 

At a high level - I think it is best to have QA focus on testing an environment that is as close to production as possible. I don't think they should be using this feature or forcing the app into offline mode. They should be turning off their Wi-Fi to test offline IMO.

**Comment by deetergp** [comment]
I agree with what @neil-marcellini and @marcaaron have said. This doesn't seem particularly useful since it won't affect production/end-users. I am going to close this one out. Maybe we'll care about it more once we're on the other side of WAQ.
